lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,488 @@
+	req := &kmspb.GetPublicKeyRequest{
+		Name: keyID.keyVersionName,
+	}
+	// Retry fetching the public key while the key is pending creation.
+	resp, err := retryWhilePending(ctx, g.clock, g.kmsClient.GetPublicKey, req)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error fetching public key")
+	}
+
+	block, _ := pem.Decode([]byte(resp.Pem))
+	if block == nil {
+		return nil, trace.BadParameter("GCP KMS key %s has invalid public key PEM", keyID.keyVersionName)
+	}
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error parsing public key pem")
+	}
+
+	return &kmsSigner{
+		ctx:       ctx,
+		kmsClient: g.kmsClient,
+		clock:     g.clock,
+		keyID:     keyID,
+		public:    pub,
+	}, nil
+}
+
+func (s *kmsSigner) Public() crypto.PublicKey {
+	return s.public
+}
+
+func (s *kmsSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	requestDigest := &kmspb.Digest{}
+	switch opts.HashFunc() {
+	case crypto.SHA256:
+		requestDigest.Digest = &kmspb.Digest_Sha256{
+			Sha256: digest,
+		}
+	case crypto.SHA512:
+		requestDigest.Digest = &kmspb.Digest_Sha512{
+			Sha512: digest,
+		}
+	default:
+		return nil, trace.BadParameter("unsupported hash func for GCP KMS signer: %v", opts.HashFunc())
+	}
+
+	resp, err := doGCPRequest(s.ctx, s.clock, s.kmsClient.AsymmetricSign, &kmspb.AsymmetricSignRequest{
+		Name:   s.keyID.keyVersionName,
+		Digest: requestDigest,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "error while attempting GCP KMS signing operation")
+	}
+	return resp.Signature, nil
+}
+
+type gcpKMSKeyID struct {
+	keyVersionName string
+}
+
+func (g gcpKMSKeyID) marshal() []byte {
+	var buf bytes.Buffer
+	buf.WriteString(gcpkmsPrefix)
+	buf.WriteString(g.keyVersionName)
+	return buf.Bytes()
+}
+
+func parseGCPKMSKeyID(key []byte) (gcpKMSKeyID, error) {
+	var keyID gcpKMSKeyID
+	if keyType(key) != types.PrivateKeyType_GCP_KMS {
+		return keyID, trace.BadParameter("unable to parse invalid GCP KMS key")
+	}
+	// strip gcpkms: prefix
+	keyID.keyVersionName = strings.TrimPrefix(string(key), gcpkmsPrefix)
+	return keyID, nil
+}
+
+type maybeRealClock struct {
+	clockwork.Clock
+	real bool
+}
+
+func newMaybeRealClock(clock clockwork.Clock, isReal bool) *maybeRealClock {
+	return &maybeRealClock{
+		Clock: clock,
+		real:  isReal,
+	}
+}
+
+func (c *maybeRealClock) isReal() bool {
+	return c.real
+}
+
+type deadlineContext struct {
+	deadline time.Time
+	context.Context
+}
+
+func (d *deadlineContext) Deadline() (time.Time, bool) {
+	return d.deadline, true
+}
+
+// withDeadline is like context.WithDeadline, but with mockable time.
+func withDeadline(ctx context.Context, clock *maybeRealClock, t time.Time) (context.Context, context.CancelFunc) {
+	// context.WithDeadline uses 1 fewer goroutines, so use that when given a
+	// real clock.
+	if clock.isReal() {
+		return context.WithDeadline(ctx, t)
+	}
```

Comment:
```suggestion
	if _, isFake := clock.(clockwork.FakeClock); !isFake {
		return context.WithDeadline(ctx, t)
	}
```

and then we can drop the extra wrapper.